framework/math/ParallelVector/ParallelVector.h

```diff
@@ -0,0 +1,164 @@
+#ifndef CHITECH_PARALLELVECTOR_H
+#define CHITECH_PARALLELVECTOR_H
+
+#include "math/chi_math.h"
+
+#include <cstdint>
+
+namespace chi_math
+{
+
+enum class VecOpType : short
+{
+  SET_VALUE = 1,
+  ADD_VALUE = 2
+};
+
+/**
+ * An abstract implementation of a parallel vector.
+ */
+class ParallelVector
+{
+public:
+  /**
+   * Initialize a parallel vector with the given local and global sizes with
+   * the given communicator whose entries are set to zero.
+   */
+  ParallelVector(uint64_t local_size,
+                 uint64_t global_size,
+                 MPI_Comm communicator);
+
+  /**Copy constructor.*/
+  ParallelVector(const ParallelVector& other);
+
+  /**Move constructor.*/
+  ParallelVector(ParallelVector&& other) noexcept;
+
+  virtual std::unique_ptr<ParallelVector> MakeCopy() const = 0;
+  virtual std::unique_ptr<ParallelVector> MakeNewVector() const = 0;
+
+  /** Returns the raw stl-vector associated with the local data of this
+   * vector.*/
+  virtual std::vector<double>& RawValues() = 0;
+
+  /// Return the size of the locally owned portion of the parallel vector.
+  uint64_t LocalSize() const { return local_size_; }
+
+  /// Return the global size of the parallel vector.
+  uint64_t GlobalSize() const { return global_size_; }
+
+  /**
+   * Read only accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double operator[](int64_t local_id) const = 0;
+
+  /**
+   * Read/write accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double& operator[](int64_t local_id) = 0;
+
+  /// Return a vector containing the locally owned data.
+  virtual std::vector<double> MakeLocalVector() = 0;
```

Comment:
Well, this goes with an alternative that I am thinking about. The make local vector business for setting the local values of a vector requires 2 copies. Once to copy out of the vector, then you use something like CopyVecToSTLVec, then you have to use Set again which does bounds checking AND another copy.

The real solution here would be to overload the `CopyValues` routine to operate on a PETSc vector.

Reply:
We should have a discussion on this

Reply:
I think we should cross the PETSc abstraction when we get to it instead of trying to build it all in before it exists. I'd rather see changes for compatibility than to try to imagine what a PETSc implementation would look like and make a judgement based on that.

Reply:
We now have `double* ParallelVector::Data()` which serves all our needs. It's pretty efficient too.